Updated doc for issue 395

achatterjee-grafana opened the pull request:
- Added new topic: "Override trace limit". 
- Updated topic "Troubleshooting" with a new section "Problem 4. Distributor is not accepting traces".

Issue: https://github.com/grafana/tempo/issues/395

joe-elliott commented on docs/tempo/website/configuration/override.md:
```diff
@@ -0,0 +1,63 @@
+weight: 30
+---
+
+# Override trace limit
```
This document opens up as if it's about one specific override and then generalizes.  I think a better layout would be:

1) List all 4 current override options.  Include details about what they do and sample log messages when they are tripped.

- ingestion_burst_size
- ingestion_rate_limit
- max_spans_per_trace
- max_traces_per_user

2) Show how to configure these using the standard overrides block:

```
overrides:
    ingestion_rate_limit:
    ...
```
3) Include details about global limit.  Global ingestion rate limit uses the distributor ring to attempt to 

```
overrides:
    ingestion_rate_strategy: global
```

4) Include details about how to do per tenant overrides:

```
overrides:
    per_tenant_override_config: /conf/overrides.yaml
```

file format:

```

annanay25 commented on docs/tempo/website/troubleshooting/_index.md:
```diff
@@ -4,10 +4,15 @@ weight: 470
 
 # Troubleshooting missing traces
-This topic helps with day zero operational issues that may come up when getting started with Tempo. It walks through debugging each part of the ingestion and query pipeline to drill down and diagnose issues. 
+This topic helps with day zero operational issues that may come up when getting started with Tempo. It walks through debugging each part of the ingestion and query pipeline to drill down and diagnose issues.
```
This page is getting a bit too long and scary, can we also split out the individual problems into smaller pages?

achatterjee-grafana commented:
Updated topic based on Joe's review. Once this is approved/ merged, I will break down the Trounshooting as you suggested @annanay25.
I need some info to complete the topic, added NEED INFO sections to add the information.

achatterjee-grafana commented:
QUESTION: What happens if there is a separate override file with per-tenant settings as well as an overrides section within the main configuration file? Which setting has the precedence? Does it mean that the overrides parameter applies to all tenants of the cluster except to the ones that have their tenant-specific overrides? Thank you for any clarification.

joe-elliott commented:
> QUESTION: What happens if there is a separate override file with per-tenant settings as well as an overrides section within the main configuration file? Which setting has the precedence? Does it mean that the overrides parameter applies to all tenants of the cluster except to the ones that have their tenant-specific overrides? Thank you for any clarification.

Per tenant settings override the settings in the main configuration file.  The main configuration file is seen as a place to put defaults.  The per tenant settings can be used to adjust to limit each tenant individually.

joe-elliott commented on docs/tempo/website/configuration/override.md:
```diff
@@ -0,0 +1,91 @@
+
+The parameters whose default values can be overridden are:
+
+   - `ingestion_burst_size` : `NEED INFO`
```
I can't say i know precisely what this means either.  It is passed here: https://pkg.go.dev/golang.org/x/time/rate#Limiter.Burst.  The `time/rate` is an implementation of a [token bucket](https://en.wikipedia.org/wiki/Token_bucket).  If you ctrl+f "burst" here:  https://cortexmetrics.io/docs/configuration/configuration-file/  you will find some descriptions of the same field in Cortex.

I think it's fine to say something like "Burst size used in span ingestion".   We currently run ingestion_burst_size == ingestion_rate_limit and I think that also would be fair guidance to provide.

joe-elliott commented on docs/tempo/website/configuration/override.md:
```diff
@@ -0,0 +1,91 @@
+   - `max_spans_per_trace` : Maximum number of spans per trace.  0 to disable. Default is `50,000`.
+   - `max_traces_per_user`: Maximum number of active traces per user, per ingester. 0 to disable. Default is `10,000`.
+
+`NEED INFO:The trigger points for each of the trace limits, and the error message to help the user identify which trace limit has been hit.`
```
`ingestion_burst_size` and `ingestion_rate_limit` both control the rate limit.  A log message would contain:
```
RATE_LIMITED: ingestion rate limit (100000 spans) exceeded while adding 10 spans
```
`max_spans_per_trace`:
```
TRACE_TOO_LARGE: totalSpans (50000) exceeded while adding 2 spans
```
`max_traces_per_user`:
```
LIVE_TRACES_EXCEEDED: max live traces per tenant exceeded: per-user traces limit (local: 10000 global: 0 actual local: 1) exceeded
```

There is also a metric that tracks this information: `tempo_discarded_spans_total`.  It has a `reason` and a `tenant` label.  The `reason` label will contain either: `trace_too_large`, `live_traces_exceeded` or `rate_limited`.  `rate_limited` applies to both the rate limit and burst size.

joe-elliott commented on docs/tempo/website/configuration/override.md:
```diff
@@ -0,0 +1,91 @@
+The parameters whose default values can be overridden are:
+
+   - `ingestion_burst_size` : `NEED INFO`
+   - `ingestion_max_batch_size` : Per-user allowed ingestion max batch size (in number of spans). Default is `1000`.
```
this needs to be removed.  no longer exists.

achatterjee-grafana commented:
Let me know when you have finished @joe-elliott. Thanks for the review.

joe-elliott commented:
@achatterjee-grafana I am done.  I believe I addressed all of your "NEED INFO" tags.

achatterjee-grafana commented:
Added changes from Joe. Will break down Troubleshooting in separate PR.

joe-elliott commented on docs/tempo/website/configuration/override.md:
```diff
@@ -0,0 +1,107 @@
+
+You can set the `overrides` to apply to all tenants within the cluster, to specific tenants, or globally to the cluster. Per tenant settings override the settings in the main configuration file. Use the main configuration file as a place to add the default values. The per tenant settings can be used to adjust to limit each tenant individually. For more information, refer to the [Override strategi ...
+
+### Standard overrides
```
Too much stuff going on in this section.  Just say:  To configure overrides add the following block to your confg yaml:

```
overrides:
  ingestion_burst_size: 50000
  ingestion_rate_limit: 50000
  max_spans_per_trace: 50000
  max_traces_per_user: 10000 
```

joe-elliott commented on docs/tempo/website/configuration/override.md:
````diff
@@ -0,0 +1,107 @@
+    ```
+## Override strategies
+
+The trace limits specified by the various parameters are, by default, applied as tenant-level limits. For example, a max_traces_per_user setting of 10000 means that each tenant within the cluster has a limit of 10000 traces per user. This is known as a `local` strategy in that the specified trace limits are local to each tenant.
````
local means "local to the distributor", not local to the tenant.  global means averaged across all distributors by using a distributor ring.

basically with a local ingestion rate of 10000 then each distributor will independently enforce 10000 spans/second in that distributor.  if you set a global rate of 10000 then it will divide 10000 by the number of distributors (say 5) and each distributor will enforce 2000 spans/second.  if you then scale up distributors to 10 then each distributor will enforce 1000 spans/second.

joe-elliott commented on docs/tempo/website/configuration/ingestion-limit.md:
````diff
@@ -0,0 +1,106 @@
+```
+LIVE_TRACES_EXCEEDED: max live traces per tenant exceeded: per-user traces limit (local: 10000 global: 0 actual local: 1) exceeded
+```
+## Specify overrides to the default settings
````
Combine the sections "Specify overrides to the default settings" and "Standard overrides".  These are basically the same thing: Telling the user how to configure non-default limits.

joe-elliott reviewed (changes requested):
One final change and we're good.  Thanks for this work!

joe-elliott commented on docs/tempo/website/configuration/ingestion-limit.md:
````diff
@@ -0,0 +1,100 @@
+1. Add the override parameters under this section. For example:
+
+```
+    overrides:
````
```suggestion
overrides:
```
